src/State/Scope.php

```diff
@@ -0,0 +1,310 @@
+     *
+     * @return $this
+     */
+    public function setFingerprint(array $fingerprint): self
+    {
+        $this->fingerprint = $fingerprint;
+
+        return $this;
+    }
+
+    /**
+     * Gets the list of strings used to dictate the deduplication of this event.
+     *
+     * @return string[]
+     */
+    public function getFingerprint(): array
+    {
+        return $this->fingerprint;
+    }
+
+    /**
+     * Sets the severity to apply to all events captured in this scope.
+     *
+     * @param null|Severity $level The severity
+     *
+     * @return $this
+     */
+    public function setLevel(?Severity $level): self
+    {
+        $this->level = $level;
+
+        return $this;
+    }
+
+    /**
+     * Gets the severity to apply to all events captured in this scope.
+     *
+     * @return null|Severity
+     */
+    public function getLevel(): ?Severity
+    {
+        return $this->level;
+    }
+
+    /**
+     * Add the given breadcrumb to the scope.
+     *
+     * @param Breadcrumb $breadcrumb     The breadcrumb to add
+     * @param int        $maxBreadcrumbs The maximum number of breadcrumbs to record
+     *
+     * @return $this
+     */
+    public function addBreadcrumb(Breadcrumb $breadcrumb, int $maxBreadcrumbs = 100): self
+    {
+        $this->breadcrumbs[] = $breadcrumb;
+        $this->breadcrumbs = \array_slice($this->breadcrumbs, -$maxBreadcrumbs);
+
+        return $this;
+    }
+
+    /**
+     * Gets the breadcrumbs.
+     *
+     * @return Breadcrumb[]
+     */
+    public function getBreadcrumbs(): array
+    {
+        return $this->breadcrumbs;
+    }
+
+    /**
+     * Adds a new event processor that will be called after {@see Scope::applyToEvent}
+     * finished its work.
+     *
+     * @param callable $eventProcessor The event processor
+     *
+     * @return $this
+     */
+    public function addEventProcessor(callable $eventProcessor): self
+    {
+        $this->eventProcessors[] = $eventProcessor;
+
+        return $this;
+    }
+
+    /**
+     * Clears the scope and resets any data it contains.
+     *
+     * @return $this
+     */
+    public function clear(): self
+    {
+        $this->tags->clear();
+        $this->extra->clear();
+        $this->user->clear();
+
+        $this->level = null;
+        $this->fingerprint = [];
+        $this->breadcrumbs = [];
+
+        return $this;
+    }
+
+    /**
+     * Applies the current context and fingerprint to the event. If the event has
+     * already some breadcrumbs on it, the ones from this scope won't get merged.
+     *
+     * @param Event $event          The event object
+     * @param int   $maxBreadcrumbs The maximum number of breadcrumbs to add to
+     *                              the event
+     *
+     * @return Event|null
+     */
+    public function applyToEvent(Event $event, int $maxBreadcrumbs = 100): ?Event
+    {
+        if (!empty($this->fingerprint) && empty($event->getFingerprint())) {
```

Comment:
To make those conditions more readable, maybe we can skip checking if the `$event` is empty? At that point it doesn't matter any more...

Reply:
The code resembles what is done in the Javascript SDK. I think that the check is there because that way the scope won't override any data already set in the event (this is valid for fingerprint and breadcrumbs as written in the docblock)

Reply:
Yes, I got that; I was suggesting that `empty($event->getFingerprint())` is enough of a check to avoid overriding existing data. The other check is to avoid setting null, but I think is superfluous and hinders readability.

Reply:
Got it now, I will change it as it is superfluous to not call the `Event::setFingerprint` method even though `$this->fingerprint` is empty. I will change it as you recommended!